test/testdrive/kafka-avro-upsert-sinks.td

```diff
@@ -0,0 +1,252 @@
+
+> CREATE SINK input_with_deletions_sink FROM input_with_deletions_keyed
+  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'input-sink' KEY (a)
+  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}' ENVELOPE UPSERT
+
+$ kafka-ingest format=avro topic=input-with-deletions schema=${schema}
+{"before": null, "after": {"row": {"a": 1, "b": 1}}}
+
+$ kafka-verify format=avro sink=materialize.public.input_with_deletions_sink sort-messages=true
+{"a": 1} {"a": 1, "b": 1}
+
+$ kafka-ingest format=avro topic=input-with-deletions schema=${schema}
+{"before": null, "after": {"row": {"a": 1, "b": 2}}}
+
+$ kafka-verify format=avro sink=materialize.public.input_with_deletions_sink sort-messages=true
+{"a": 1} {"a": 1, "b": 2}
+
+# deletion of the "shadowed" input should not cause downstream updates
+
+$ kafka-ingest format=avro topic=input-with-deletions schema=${schema}
+{"before": {"row": {"a": 1, "b": 1}}, "after": null}
+
+$ kafka-verify format=avro sink=materialize.public.input_with_deletions_sink sort-messages=true
+
+$ kafka-ingest format=avro topic=input-with-deletions schema=${schema}
+{"before": {"row": {"a": 1, "b": 2}}, "after": null}
+
+# now we should see a NULL update on the key, which means a DELETE
+
+$ kafka-verify format=avro sink=materialize.public.input_with_deletions_sink sort-messages=true
+{"a": 1}
+
+$ kafka-ingest format=avro topic=input-with-deletions schema=${schema}
+{"before": null, "after": {"row": {"a": 1, "b": 1}}}
+{"before": null, "after": {"row": {"a": 1, "b": 2}}}
+
+$ kafka-verify format=avro sink=materialize.public.input_with_deletions_sink sort-messages=true
+{"a": 1} {"a": 1, "b": 2}
+
+$ kafka-ingest format=avro topic=input-with-deletions schema=${schema}
+{"before": {"row": {"a": 1, "b": 2}}, "after": null}
+
+# removing the occluding input should "reveal" the previous input again
+#
+$ kafka-verify format=avro sink=materialize.public.input_with_deletions_sink sort-messages=true
+{"a": 1} {"a": 1, "b": 1}
+
+# Bad upsert keys
+
+! CREATE SINK invalid_key FROM input
+  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'input-sink' KEY (a)
+  WITH (consistency = true) FORMAT AVRO
+  USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}' ENVELOPE UPSERT
+Invalid upsert key: (a), there are no valid keys
+
+! CREATE SINK another_invalid_key FROM input_keyed
+  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'input-sink' KEY (b)
+  WITH (consistency = true) FORMAT AVRO
+  USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}' ENVELOPE UPSERT
+Invalid upsert key: (b), valid keys are: (a)
+
+> CREATE VIEW input_keyed_ab AS SELECT a, b FROM input GROUP BY a, b
+
+! CREATE SINK invalid_sub_key FROM input_keyed_ab
+  INTO KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'input-sink' KEY (a)
+  WITH (consistency = true) FORMAT AVRO
+  USING CONFLUENT SCHEMA REGISTRY '${testdrive.schema-registry-url}' ENVELOPE UPSERT
+Invalid upsert key: (a), valid keys are: (a, b)
```

Comment:
Thank you for the tests! If you can, can you please add some for multi-part keys, that is, a source has a multi-part key and the sink should fail if the requested key is not strictly identical to the source key (so not a prefix of, a subset/superset or permutation thereof). That would be much appreciated!

Reply:
You mean to use an actual source in addition to using a `VIEW` as I do here? I can do that, for sure!

Regarding keys, we explicitly allows specifying a key `(a, b)` for the sink when the collection is keyed only by `(a)` because that will also be a unique key. And in this test right here we test that `(a)` is not accepted when the unique key is `(a, b)`. Should I add a test to verify that `(b)` is also rejected?

Reply:
Yes, I think a test with a multi-part source would be nice. Multi-part keys of any kind are usually underrepresented in tests  but are actually seen in real life (with up to 5  parts nonetheless!) so lets add them whenever possible.

Yes you are right, (a, b) should be accepted. If you can add a test for (b) that would be much appreciated.